### PR TITLE
fix(test): Throw if elements are not found

### PR DESCRIPTION
**Pull request by lasjorg**
#### Pre-Submission Checklist

- [x] Your pull request targets the `master` branch of freeCodeCamp/testable-projects-fcc
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] Your changes have been tested either locally or using a newly created CDN based on your fork's testable-projects-fcc/build/bundle.js file

#### Type of Change

- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [x] Tested changes locally.

- [ ] Closes currently open issue: Closes #XXX

#### Description

Right now tests 6 and 7 will pass with no code for the "Build a Random Quote Machine" project. Not sure what the best way of handling this is but I think just throwing in an else block if the elements are not found should work just fine.

Forum post: https://forum.freecodecamp.org/t/build-a-random-quote-machine-bug-in-test/472095

### CODE REVIEW

File: src/project-tests/quote-machine-tests.js [hunk 1]
```diff
@@ -96,6 +96,8 @@ export default function createRandomQuoteMachineTests() {
               }, 500);
             });
           });
+        } else {
+          assert.fail('element with id="text" not found');
```
Comment (ShaunSHamilton) [inline src/project-tests/quote-machine-tests.js:100]: Personally, I would have gone with asserting on the element we are testing:
```js
assert.exists(text);
```
Same goes for other tests. Thoughts?

Comment (lasjorg) [reply src/project-tests/quote-machine-tests.js:100]: Shouldn't it be `notExists` then as we are expecting it to be falsy in the else block?

Anyway, I guess it's more explicit so I'm fine with changing it.

Comment (ShaunSHamilton) [reply src/project-tests/quote-machine-tests.js:100]: To clarify, there should be no `else` clause here - we should assert that the element exists. The test **should** fail, if it does not exist.

Comment (lasjorg) [reply src/project-tests/quote-machine-tests.js]: We already have tests for the elements, tests 2 and 3. The two tests (6 and 7) are testing for the text content of the element but never fails if the elements are null.

File: src/project-tests/quote-machine-tests.js [hunk 2]
```diff
@@ -96,6 +96,8 @@ export default function createRandomQuoteMachineTests() {
               }, 500);
             });
           });
+        } else {
+          assert.exists(text, 'element with id="text" not found');
         }
```
Comment (ShaunSHamilton) [inline src/project-tests/quote-machine-tests.js:101]:
```suggestion
        }
        assert.exists(text, 'element with id="text" not found');
```

File: src/project-tests/quote-machine-tests.js [hunk 3]
```diff
@@ -117,6 +119,8 @@ export default function createRandomQuoteMachineTests() {
               }, 500);
             });
           });
+        } else {
+          assert.exists(author, 'element with id="author" not found');
         }
```
Comment (ShaunSHamilton) [inline src/project-tests/quote-machine-tests.js:124]:
```suggestion
        }
        assert.exists(author, 'element with id="author" not found');
```

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
This is more what I mean. I am just a fan of less code 🤷‍♂️

File: src/project-tests/quote-machine-tests.js (hunk 1 above)
Comment (ShaunSHamilton) [reply src/project-tests/quote-machine-tests.js]: Also, a user brought up a good point - we should add this test to all that do not have it: https://forum.freecodecamp.org/t/build-a-random-quote-machine-bug-in-test/472095/3

**Comment by lasjorg** [comment]
@ShaunSHamilton Sorry for the lack of update on this.

I had it working to where the example project and the empty project passed and failed respectively as needed. But then I tested with a React version of the project I had laying around and it fails. I rolled back to the previous code and it still fails some of the tests. It seems [this PR](https://github.com/freeCodeCamp/testable-projects-fcc/pull/1253) broke it because the project passes with the old test code. I get the feeling the React conditional rendering I have in the project doesn't work with the new code.

I might play around with it some more at some point, but right now I don't really have anything.

**Comment by ShaunSHamilton** [comment]
@lasjorg The only thing that comes to my mind that could be an issue with the linked PR's changes is if the script is added before the body of the HTML.

In your old React form of the project, how are you adding the test suite?

**Comment by lasjorg** [comment]
@ShaunSHamilton No that's not it. As I said, it still passes with the old test code from before that PR.

Here is a replit, you can switch the test scripts in the HTML file (current and old).
https://replit.com/@lasjorg/random-quote-tests

**Comment by ShaunSHamilton** [comment]
Just to mention it, if you were unaware: Removing the `hasLoaded` in your app allows it to pass the tests:
```diff
return (
    <>
-       {quotes && hasLoaded && (
+      {quotes && (
        <div id="quote-box" className={`App ${hasLoaded && 'visible'}`}>
          <Quote title={quotes.author} content={quotes.content} />
          <Buttons
            handleClick={handleClick}
            disabled={!hasLoaded}
            href={getTweetLink(quotes.author, quotes.content)}
          />
        </div>
      )}
    </>
  );
```

_I also renamed `onClick` as a prop, because it was confusing me_ 🤫

**Comment by lasjorg** [comment]
Right but as I said it's just some code that fails with the current test that didn't with the old. It wasn't about changing the code to make it pass it was just an example of code that used to pass that now fails.

I will just push what I have now because I started to mess around with it and I'm losing track of what I have tested. I added a test for the width of the container to not allow for full-width containers counting as centered. The current code should pass the example project, fail the empty project, and pass the React project with the change to the conditional code as you suggested.

Having all that code before the promises just feels off to me, anyway it's what I have got so far so I might as well just push it.

**Review by ShaunSHamilton — COMMENTED** [review]
This still causes your example app to fail, unless the `hasLoaded` logic is removed from the render  in `App.js`.

So, I am fine with just going ahead with this for better assertion messages. Is that what we want?

**Review by ShaunSHamilton — APPROVED** [review]
Thank you, for baring with me, @lasjorg .

I am happy with what this accomplishes.

---

Something for the next reviewer: Is there any reason the `25px` leeway is too little for the `testHorizontallyCentered` function?

File: src/project-tests/quote-machine-tests.js [hunk 4]
```diff
@@ -123,12 +139,21 @@ export default function createRandomQuoteMachineTests() {
       it(`When the #new-quote button is clicked, my quote machine
       should fetch a new quote and display it in the #text element.`, function () {
         let prevText;
+
         const newQuoteBtn = document.getElementById('new-quote');
+        newQuoteBtn.click();
```
Comment (moT01) [inline src/project-tests/quote-machine-tests.js:144]: I think this needs to be moved below the `prevText = ...` (line 151) - or else `prevText` and `newText` would always be the same no?

File: src/project-tests/quote-machine-tests.js [hunk 5]
```diff
@@ -151,12 +176,21 @@ export default function createRandomQuoteMachineTests() {
       it(`My quote machine should fetch the new quote's author when
       the #new-quote button is clicked and display it in the #author element.`, function () {
         let prevAuth;
+
         const newQuoteBtn = document.getElementById('new-quote');
+        newQuoteBtn.click();
```
Comment (moT01) [inline src/project-tests/quote-machine-tests.js:181]: Same as other suggestion here

**Review by moT01 — COMMENTED** [review]
I was struggling with testing this - finally figured it out. I made two suggestions.

**Comment by lasjorg** [comment]
@moT01 I forgot all about this PR. Are there any other issues that need to be addressed here?

---

A forum post that made me remember this PR. This project can be passed with an almost empty project as it is now.

https://forum.freecodecamp.org/t/im-getting-12-12-tests-passed-on-random-quote-machine-but-ive-barely-written-7-lines-of-code/494900

**Review by moT01 — APPROVED** [review]
LGTM @lasjorg 🎉 Sorry for the delay here - Thanks for the work on this and all your contributions 🎉

**Pull request merged**